File: Python/textbook_track/chapter09/ba9m.py
# Implement BetterBWMatching
import functools as ft
import sys

from collections import Counter


def calc_first_occurrences(transformed_string):
    letter_counts = Counter(transformed_string)
    first_occurrence = {}
    ix = 0
    for letter in sorted(letter_counts.keys()):
        first_occurrence[letter] = ix
        ix += letter_counts[letter]
    return first_occurrence


def calc_count_matrix(transformed_string, unique_characters):
    length = len(transformed_string)
    count_matrix = {letter: [0] * (length+1) for letter in unique_characters}
    for ix, current_letter in enumerate(transformed_string):
        for char in unique_characters:
            counts = count_matrix[char]
            counts[ix+1] = counts[ix] + 1 if current_letter == char else counts[ix]
    return count_matrix


def find_pattern_in_text(pattern, first_occurrences, count_matrix, length):
    top_pointer = 0
    bottom_pointer = length - 1
    while pattern:
        letter = pattern.pop()
        if letter in first_occurrences:
            top_pointer = first_occurrences[letter] + count_matrix[letter][top_pointer]
            bottom_pointer = first_occurrences[letter] + count_matrix[letter][bottom_pointer+1] - 1
        else:
            return 0

        if top_pointer > bottom_pointer:
            return 0

    return bottom_pointer - top_pointer + 1


def perform_pattern_matching(transformed_text, patterns):
    length = len(transformed_text)
    first_occurrences = calc_first_occurrences(transformed_text)
    unique_letters = frozenset(first_occurrences.keys())
    count_matrix = calc_count_matrix(transformed_text, unique_letters)
    count_matches = ft.partial(find_pattern_in_text, first_occurrences=first_occurrences, count_matrix=count_matrix,
                               length=length)
    return [count_matches(list(pattern)) for pattern in patterns]


def main():
    reader = sys.stdin
    transformed_text = next(reader).rstrip()
    patterns = next(reader).split()
    result = perform_pattern_matching(transformed_text, patterns)
    print(' '.join(map(str, result)))


if __name__ == "__main__":
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9M.scala
package TextbookTrack.Chapter09

object BA9M {
  import scala.annotation.tailrec

  private def calcFirstOccurrencePositions(text: List[Char]): Map[Char, Int] = {
    val letterCounts: Map[Char, Int] = text.groupMapReduce(identity)(_ => 1)(_ + _)
    letterCounts
      .keysIterator
      .toList
      .sorted
      .foldLeft((Map.empty[Char, Int], 0)) {
        case ((firstOccurrence, ix), letter) => (firstOccurrence.updated(letter, ix), ix + letterCounts(letter))
      }
      ._1
  }

  private def calcCountMatrix(transformedText: List[Char], n: Int, uniqueLetters: Set[Char]): Map[Char, Array[Int]] = {
    val countMatrix: Map[Char, Array[Int]] = uniqueLetters.map(_ -> Array.fill(n + 1)(0)).toMap
    for {
      (letter, ix) <- transformedText.view.zipWithIndex
      char <- uniqueLetters
    } {
      val counts: Array[Int] = countMatrix(char)
      counts(ix + 1) = if (letter == char) counts(ix) + 1 else counts(ix)
    }

    countMatrix
  }

  private def patternMatching(pattern: String,
                              firstOccurrences: Map[Char, Int],
                              countMatrix: Map[Char, Array[Int]],
                              n: Int): Int = {
    val reversedPattern: Iterator[Char] = pattern.reverseIterator

    @tailrec
    def loop(topPointer: Int, bottomPointer: Int): Int = reversedPattern.nextOption() match {
      case None => bottomPointer - topPointer + 1
      case Some(letter) =>
        val updatedPointers: Option[(Int, Int)] =
          firstOccurrences
            .get(letter)
            .map{ letterOccurrence =>
              val letterCounter: Array[Int] = countMatrix(letter)
              (letterOccurrence + letterCounter(topPointer), letterOccurrence + letterCounter(bottomPointer + 1) - 1)
            }
        updatedPointers match {
          case None => 0
          case Some((top, bottom)) => if (top > bottom) 0 else loop(top, bottom)
        }
    }

    loop(0, n - 1)
  }

  def performPatternMatching(transformedText: List[Char], patterns: List[String]): List[Int] = {
    val length: Int = transformedText.length
    val firstOccurrences: Map[Char, Int] = calcFirstOccurrencePositions(transformedText)
    val uniqueLetters: Set[Char] = firstOccurrences.keySet
    val countMatrix: Map[Char, Array[Int]] = calcCountMatrix(transformedText, length, uniqueLetters)
    patterns.map(patternMatching(_, firstOccurrences, countMatrix, length))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val transformedText: List[Char] = reader.next().toList
    val patterns: List[String] = reader.next().split(" ").toList
    val result: List[Int] = performPatternMatching(transformedText, patterns)
    println(result.mkString(" "))
  }
}
